### ISSUE TITLE
Are we using /32 + gateway with DHCP?

### ISSUE THREAD

**Issue by fasaxc** [Metaswitch/calico#186]
In https://github.com/Metaswitch/calico/pull/183#issuecomment-73926617 @Lukasa said that we're  sending /32s with a gateway in DHCP.  I think that is a misconfiguration.  shouldn't we be using the pool IP subnet?

To clarify: I think DHCP requires that the gateway is within the advertised subnet, so it can't be a /32.

**Comment by liljenstolpe** [comment]
We should be.  That was the original design.  A /32 (and out of network gateway) is an undefined behavior in the DHCP RFC's.  I really wanted to play this game, but not every OS supports (or they do it differently).  The design was we push the network (i.e. 10.65..8.9/14, gw 10.64.0.1) to the guest via DHCP.

**Comment by Lukasa** [comment]
To be even clearer, we don't control DHCP currently: we just let OpenStack do it. =)

**Comment by neiljerram** [comment]
I'm confused. Normal DHCP gives a /32 IP address to a host on a network. How is what Calico (or OpenStack) is doing different from that?

**Comment by fasaxc** [comment]
I think we're talking about different things.  DHCP is supposed to assign a single IP address to a host.  sometimes we call that a /32, but that's not what I'm referring to.  DHCP is also expected to assign a subnet mask to tell the client what subnet its IP address belongs to.  So DHCP would assign IP 10.1.2.3/16, for example to tell the client that its single IP address is 10.1.2.3 but its subnet is 10.1.0.0/16. 

AFAIK, DHCP can't assign multiple IP addresses or a whole subnet to an individual host.

**Comment by liljenstolpe** [comment]
That's very wrong - we must be doing something odd.  In normal openstack, the guest gets an address in it's subnet/network.  Since we are running shared network (we are, aren't we?), then it would get an address out of whatever network the guest is assigned to.

DHCP assigns an address out of a subnet (i.e. 10.65.8.9/14).  It does NOT assign a /32.  If we are doing that, then we've configured something wrong in openstack (or openstack is completely broken).  This MAY work with modern Linux (obviously it does).  I am pretty sure it's going to bork other guest OS's.

**Comment by neiljerram** [comment]
I could be wrong, but I don't think Cory (in his comment that triggered this issue) meant /32 in the sense that Shaun understood. I'm pretty sure that Calico + OpenStack behaves as you would want, Chris: i.e. it distributes an IPv4 address and a subnet prefix with exactly the prefix length as configured in the OpenStack subnet config.

**Comment by Lukasa** [comment]
Yup. Pretty sure I very confidently answered a different question than Shaun asked. My apologies folks, that misunderstanding is on me.

**Comment by neiljerram** [comment]
OK, I've just double checked.  Here's what ifconfig and route -n say on the guest in a Calico cluster:

    ubuntu@nj-ubuntu:~$ ifconfig
    eth0      Link encap:Ethernet  HWaddr fa:16:3e:5b:2c:de
              inet addr:192.168.0.71  Bcast:192.168.0.255  Mask:255.255.255.0
              inet6 addr: fe80::f816:3eff:fe5b:2cde/64 Scope:Link
              UP BROADCAST RUNNING MULTICAST  MTU:1400  Metric:1
    ...
    
    ubuntu@nj-ubuntu:~$ route -n
    Kernel IP routing table
    Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
    0.0.0.0         192.168.0.1     0.0.0.0         UG    0      0        0 eth0
    192.168.0.0     0.0.0.0         255.255.255.0   U     0      0        0 eth0

In other words, the guest is seeing a /24 subnet, as it should.
